### PR TITLE
Add rich nav specific pipeline to enable Rich Navigation

### PR DESCRIPTION
**Pull request by jepetty**
I've added a new Rich Navigation specific build (which will appear as a new build pipeline), indexing on every push to master. This will allow you to use Rich Navigation in Visual Studio to enable cross-repository searches, serverless navigation at https://online.visualstudio.com/github/dotnet/msbuild, and integrated GitHub C# navigation once that is released (or, comment below if you want to be added to the preview now 😊).

## Rich Code Navigation

[Docs Link](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/1047/Rich-Code-Navigation)
With Rich Code Navigation, you can use editor-level navigation capabilities (like peek definition, find all references, and even diagnostics) on a pull request, without requiring a local checkout. This is available across all files and dependencies of your repository.

Rich Code Navigation is available for Visual Studio (for repos hosted on GitHub or Azure Repos) and Visual Studio Code (for repos hosted on GitHub).

Rich Code Navigation supports these capabilities:

- Hover
- Peek definition and Go to definition
- Peek references and Find all references
- Go to type definition
- Go to implementation
- Diagnostics

### CODE REVIEW

**Comment by jepetty** [comment]
I've temporarily enabled this on one of your normal CI builds just to get a baseline view of how long this task will take - I'll remove this addition once I get a successful build.

FWIW I've added this to a handful of dotnet repos over the last week, and runtime has by far the most complex build architecture, so any feedback/guidance is greatly appreciated :)

File: eng/common/templates/job/job.yml
Comment (safern) [inline eng/common/templates/job/job.yml:135]: All the changes under `eng/common/` need to be done in dotnet/arcade repository directly if not they will be overridden by our bot.

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:34]: Why are we only building coreclr? Most of the C# code is built in libraries subset.

File: eng/pipelines/coreclr/templates/build-job.yml
Comment (safern) [inline eng/pipelines/coreclr/templates/build-job.yml:40]: Instead of putting this here flow the parameter as a jobParameter in the entrypoint yml, `rich-nav.yml`

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:35]: osGroup, targetRid, and archGroup are not needed here, those are calculated based on the Platform when you reference `platform-matrix.yml`

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:32]: Use `platform-matrix.yml` template instead as we do elsewhere.

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:40]: Once you use `platform-matrix.yml` the build config will need to be a parameter and not a jobParameter.

**Review by safern — CHANGES_REQUESTED** [review]

**Comment by safern** [comment]
Thanks @jepetty! I've left some comments.

**Comment by ViktorHofer** [comment]
@jepetty is native code supported as well for this?

File: eng/common/templates/job/job.yml
Comment (jepetty) [reply eng/common/templates/job/job.yml:135]: Definitely! Sorry, I should have linked to [this PR](https://github.com/dotnet/arcade/pull/6345) - I only included them here as well to test.

File: eng/pipelines/runtime-richnav.yml
Comment (jepetty) [reply eng/pipelines/runtime-richnav.yml:34]: Thanks so much for the question! Apologies for my naivety, I was confused about which builds should be included. Is there any other build I should be including? Also I mostly picked the architecture at random (we need to run on Windows) - does this look like an appropriate choice to you?

File: eng/pipelines/runtime-richnav.yml
Comment (jepetty) [reply eng/pipelines/runtime-richnav.yml:32]: Thanks! I'm switching to that with just "Windows_NT_X64" as the target platform - is there value in adding Windows_NT_x86 or others?

File: eng/pipelines/coreclr/templates/build-job.yml
Comment (jepetty) [reply eng/pipelines/coreclr/templates/build-job.yml]: Good call! I changed that as well

**Comment by jepetty** [comment]
Couple of quick comments for reviewers:
1. Because of the complexity of the build system, I'd like to wait to merge this until my fixes to arcade are in and I can build a one-off build of runtime-richnav.yml to verify everything works as expected
1. We currently support C++ (though potentially only through vcxproj, I'd need to double check with the C++ team), C#/VB, TypeScript/JavaScript, and Java. Once my improvements to arcade are in, I'll be able to pass the languages as a parameter to the Rich Nav task as well.

Thanks for the help and feedback!

File: eng/common/templates/job/job.yml
Comment (jepetty) [reply eng/common/templates/job/job.yml]: I went ahead and removed this - there's enough things to verify with your pipelines that I think it's worth just waiting for the official merge

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:32]: Let's remove `-runtimeConfiguration` from both of these guys... debug for all is fine.

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:26]: Make this Debug instead.

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:41]: we can remove `testGroup`

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:42]:
```suggestion
        nameSuffix: Mono
```

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:12]: do we need `TeamName`?

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [inline eng/pipelines/runtime-richnav.yml:5]: I think `dev/infrastructure` is not a very common branch to navigate on, so might make sense to not include it.

File: eng/pipelines/runtime-richnav.yml
Comment (safern) [reply eng/pipelines/runtime-richnav.yml:32]: Also, are we going to pass down the MSBuild arg to restore the package?

`/p:EnableRichCodeNavigation=true`?

We could do that directly in `global-build-job.yml` as we do for other things.
https://github.com/dotnet/runtime/blob/7601d44ca06fc977c4a3c7a33c1fc864f8f5ad78/eng/pipelines/common/global-build-job.yml#L53-L57

we could add something like:
```yml
      - name: _richCodeNavigationParam
        ${{ if eq(parameters.enableRichCodeNavigation, true) }}:
          value: /p:EnableRichCodeNavigation=true
        ${{ if ne(parameters.enableRichCodeNavigation, true) }}:
          value: ''
```

and then flow it here: https://github.com/dotnet/runtime/blob/7601d44ca06fc977c4a3c7a33c1fc864f8f5ad78/eng/pipelines/common/global-build-job.yml#L78

File: eng/pipelines/common/global-build-job.yml
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:34]: let's default this parameter to false in this template by adding it to the parameters list and then do a boolean comparison:

```yml
${{ if eq(parameters.enableRichCodeNavigation, true) }}:
```

File: eng/pipelines/common/global-build-job.yml
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:36]: should we allow this to be passed as a parameter when referencing global-build-job? For example the `libs.refs+libs.src` job is not enabling it for cpp yet.

File: eng/pipelines/runtime-richnav.yml
Comment (jepetty) [reply eng/pipelines/runtime-richnav.yml:32]: Ah yes. This is done in the [template](https://github.com/dotnet/runtime/blob/master/eng/common/templates/job/job.yml#L68..L70), but I agree with you, for repro-ability/readability I think we should pipe it through as well. Thanks for the detailed description, I'll make this change!

File: eng/pipelines/common/global-build-job.yml
Comment (jepetty) [reply eng/pipelines/common/global-build-job.yml:34]: Great idea! Done.

File: eng/pipelines/common/global-build-job.yml
Comment (jepetty) [reply eng/pipelines/common/global-build-job.yml:36]: Oh great point, good catch! I'll update, thanks

File: eng/pipelines/runtime-richnav.yml
Comment (jepetty) [reply eng/pipelines/runtime-richnav.yml:12]: Nope, you're right. Good catch! Removing

**Review by safern — APPROVED** [review]
LGTM! We just need to wait for the arcade PR to make it this way 😄 

What is the way to test if this is good enough?

**Comment by jepetty** [comment]
Please temporarily ignore my last hack-tacular commit that replaces the regular CI with the Rich Nav CI. I'm doing this purely for testing, and will revert before merging.

**Comment by jepetty** [comment]
It looks like the build timed out: https://dev.azure.com/dnceng/public/_build/results?buildId=861300&view=logs&j=b944db00-7f1c-5e05-ff25-7ef34c725b70&t=69de74b2-f6ff-5dd9-e23f-04c7ae7fa203. We're actively working to bring our build time down - would we want to increase the timeout of this build or wait for our improvements to be made?

**Comment by safern** [comment]
>  We're actively working to bring our build time down - would we want to increase the timeout of this build or wait for our improvements to be made?

Given that it will be a separate pipeline... maybe running batched or on a scheduled, I think it is OK to take a relatively long build by increasing timeout and just wait for the improvements while we're already running.

**Comment by safern** [comment]
@jepetty do you have any update on this PR? Is it ready or are you still working on it? If so, can we close it on the meantime as it seems pretty stale at the moment?

**Comment by jepetty** [comment]
@safern thanks for checking! We had some infrastructure changes on our side, and I've confirmed the pipeline now completes [successfully](https://dev.azure.com/dnceng/public/_build/results?buildId=974471&view=logs&j=f73b9f71-ec01-5f8c-7657-d0a24e86be8b&t=4002bf64-df1f-5eba-cab7-c27c03fae20b). This can be merged whenever you're ready, and then I'll create a new pipeline for runtime-richnav!

**Review by safern — APPROVED** [review]
Looks good. Thank you!

**Comment by safern** [comment]
@jepetty what are the next steps once the new pipeline is added? We can adjust the pipeline accordingly if something is not working but this looks great as an initial step.

How can we test that this is working fine?

Also, can we add support for native code at some point?

Should we index test projects?

**Pull request merged**
